Add possibility of modifying dynamic objects behavior

forkerer opened the pull request:
This pull request aims to add possibility of modifying behavior of dynamic objects. This includes properties like their mass, particle effects played by them, if their explode or not, etc. By now mta has setObjectProperty() function implemented, but it's limited to custom created objects, and doesn't allow to change behavior of original world objects. All default properties, along with simple documentation, can be found in object.dat in GTA SA data folder. All changed properties are restored to default once user leaves server.

# 1. Property groups used by certain models IDs
SA by default contains 160 property groups for dynamic objects, multiple models can share single properties group, i've added possibility to change which group is used by which model, which required addition of those 3 functions:
```lua
-- Returns properties group used by given model ID, will return 0 if model isn't dynamic, as it's defaulted to that in engine
engineGetModelPhysicalPropertiesGroup(modelID) 

-- Changes which property group given model will use
engineSetModelPhysicalPropertiesGroup(modelID, groupID)

-- Restores model to default state
engineRestoreModelPhysicalPropertiesGroup(modelID)
```
I didn't notice any change when i've changed group id's of static objects (or any non dynamic stuff in CModelInfo table), but game is stable when those are changed for them, so i didn't disable ability to do so.

# 2. Properties themselves
There are multiple adjustable properties for each group, most of the changes made in them are visible after object is created, so most changes require you to get far away from object, and back to become visible.
Those 3 functions were created to give ability to modify those properties:
```lua
-- Property setters
engineSetObjectGroupPhysicalProperty(groupID, objectgroup-modifiable property, [...args] )

-- Property getters
engineGetObjectGroupPhysicalProperty(groupID, objectgroup-modifiable property)

-- Used to restore default property values
engineRestoreObjectGroupPhysicalProperties(groupID)
```

Those are the modifiable properties, along with their lua names and types:
```cpp
enum eObjectGroup::Modifiable
{
    MASS,                //mass                         float
    TURNMASS,            //turn_mass                    float
    AIRRESISTANCE,       //air_resistance               float
    ELASTICITY,          //elasticity                   float
    BUOYANCY,            //buoyancy                     float
    UPROOTLIMIT,         //uproot_limit                 float
    COLDAMAGEMULTIPLIER, //col_damage_multiplier        float
    COLDAMAGEEFFECT,     //col_damage_effect            eObjectGroup::DamageEffect
    SPECIALCOLRESPONSE,  //special_col_response         eObjectGroup::CollisionResponse
    CAMERAAVOID,         //avoid_camera                 bool
    EXPLOSION,           //cause_explosion              bool
    FXTYPE,              //fx_type                      eObjectGroup::FxType
    FXOFFSET,            //fx_offset                    CVector
    FXSYSTEM,            //fx_system                    takes string, uses FxSystemBlueprint
    SMASHMULTIPLIER,     //smash_multiplier             float
    BREAKVELOCITY,       //break_velocity               CVector
    BREAKVELOCITYRAND,   //break_velocity_randomness    float
    BREAKMODE,           //break_mode                   eObjectGroup::BreakMode
    SPARKSONIMPACT       //sparks_on_impact             bool
};
```
The specialized properties enums looks like so:
```c
enum DamageEffect // Dictates which damage effect is applied to object on collision
{
    NO_EFFECT = 0,          //none              Object doesn't change at all once it's damaged
    CHANGE_MODEL = 1,       //change_model      Some of the objects change model on collision, those use this
    SMASH_COMPLETELY = 20,  //smash             Object is smashed
    CHANGE_THEN_SMASH = 21, //change_smash      First CHANGE_MODEL, afterwards smash on collision
    BREAKABLE = 200,        //breakable         Object is breakable normally
    BREAKABLE_REMOVED = 202 //breakable_remove  object.dat says: '(ie. never regenerated after destroyed)'
};

enum CollisionResponse // How object responds to being collided with
{
    NO_RESPONSE, //none
    LAMPPOST,    //lamppost
    SMALLBOX,    //small_box
    BIGBOX,      //big_box
    FENCEPART,   //fence_part
    GRENADE,     //grenade
    SWINGDOOR,   //swingdoor
    LOCKDOOR,    //lockdoor
    HANGING,     //hanging
    POOLBALL     //poolball
};

enum FxType // Dictates when particles will be created when colliding with object
{
    NO_FX,                //none
    PLAY_ON_HIT,          //play_on_hit
    PLAY_ON_DESTROYED,    //play_on_destroyed
    PLAY_ON_HIT_DESTROYED //play_on_hitdestroyed
};

enum BreakMode // Dictates how object can be damaged, info from object.dat:
{
    NOT_BY_GUN, //not_by_gun    'not breakable by gun'
    BY_GUN,     //by_gun        'breakable'
    SMASHABLE,  //smashable     'smashable'
};
```
### Fx system:
Any particle system included in SA could be used by dynamic objects, but after testing i noticed that looping particle systems stay on map forever (even after leaving server), so i had to disable ability to use those, only non looping system are acceptable by fx_system property, those are: (from effects.fxp file)
```
wallbust,
shootlight,
puke,
explosion_door,
explosion_crate,
explosion_barrel,
blood_heli,
tree_hit_palm,
tree_hit_fir,
water_swim,
water_splsh_sml,
water_splash_big,
water_splash,
water_hydrant,
tank_fire,
riot_smoke,
gunsmoke,
gunflash,
explosion_tiny,
explosion_small,
explosion_molotov,
explosion_medium,
explosion_large,
explosion_fuel_car,
exhale,
camflash,
prt_wake
```

# 3. Test resource
DOWNLOAD: [props.zip](https://github.com/multitheftauto/mtasa-blue/files/2776563/props.zip)
I'm including simple resource containing editor for those properties, that looks like that:

![image](https://user-images.githubusercontent.com/923242/52452890-d7333400-2b3c-11e9-9c32-b5c04ccfaff6.png)

Simply start resource, and write /showgui to make it visible, same command to hide it back again.

Properties editor:
-- Changes properties of given group
Model editor:
-- Gives ability to change properties group used by given model
Presets:
-- Some simple presets, and ability to restore properties of all groups.
-- Given presets are: make everything explode, make everything easy to destroy, make everything undestroyable

Commit:
Merge remote-tracking branch 'multitheftauto/master'

Commit:
Merge remote-tracking branch 'multitheftauto/master'

Commit:
Merge remote-tracking branch 'multitheftauto/master'

Commit:
Merge remote-tracking branch 'multitheftauto/master'

Commit:
Basic implementation of required classes

Commit:
Possibility to get object properties group from CModelInfo

Commit:
Ability to change model id -> object properties group relations

Commit:
enum definition

Commit:
moved function to the end of CGame*, small cleanup

Commit:
Fixed issues with not starting, a lot of changes

- changed the way of how backup ObjectDynamicInfo works
- added functions to get/set/restore model dynamic properties group
- some other small changes

Commit:
Basic functions, starting renaming

Commit:
The giant renaming

Commit:
model update

Commit:
model/naming changes, rest of setters from lua.

Commit:
naming, code shortening thanks to botder

Commit:
Enums cleanup, to their own namespace they go

also added missing property change

Commit:
Added missing restore function, naming changes

Commit:
enums namespace change

Commit:
Ability to get properties from lua

Commit:
enum memory layout issues fix

Commit:
Added ability to change fx system used by object group

Commit:
Additional safety checks in lua functions

Commit:
Restore object data changes, refactored lua methods

Commit:
adjust properties count to real 160, instead of 255,

Commit:
Ability to restore default group properties from lua

Commit:
Enum class name change

Commit:
Stuff i've noticed when making pull request

Commit:
Moved stuff

Commit:
Formatting fix

Einheit-101 commented:
avoid_camera is interesting, is it possible to set this property to projectile objects? Because its highly annoying that projectiles interact with the camera.

qaisjp commented:
Lua API seems reasonable. Is it possible to create additional object groups or is it limited to the default set of groups?

forkerer commented:
> Lua API seems reasonable. Is it possible to create additional object groups or is it limited to the default set of groups?

Limited to default 160 groups, dunno how many references to current table location are in the gta sa, and current amount is plenty for what's needed i think.

Commit:
Merge latest master, resolve conflicts

patrikjuvonen commented on Client/game_sa/CObjectGroupPhysicalPropertiesSA.cpp:
```diff
@@ -0,0 +1,372 @@
+CObjectGroupPhysicalPropertiesSA::CObjectGroupPhysicalPropertiesSA()
+{
+    m_pInterface = nullptr;
+}
```
```cpp
CObjectGroupPhysicalPropertiesSA::CObjectGroupPhysicalPropertiesSA() : m_pInterface(nullptr) {}
```
Prefer initializer list here instead [[1]](https://www.geeksforgeeks.org/when-do-we-use-initializer-list-in-c/)

patrikjuvonen commented on Client/game_sa/CObjectGroupPhysicalPropertiesSA.cpp:
```diff
@@ -0,0 +1,372 @@
+
+float CObjectGroupPhysicalPropertiesSA::GetAirResistance()
+{
+    ChangeSafeguard();
```
1. Why is `ChangeSafeguard();` called here, but not consistently across all getters?
2. Maybe also change all applicable member functions to `const`? [[1]](https://www.geeksforgeeks.org/const-member-functions-c/)

patrikjuvonen committed:
Clean up project comments

Commit:
Changes recommended in code review

1. Added const modfier to functions where possible
2. Changed constructor to use initializer list
3. Removed unnecessary call to ChangeSafeguard() from one of the getters

patrikjuvonen reviewed (approved):
Thanks for the changes. Looks good to me!

I'm looking at this line: https://github.com/multitheftauto/mtasa-blue/pull/784/files?w=1#diff-72a39fa83b01cb9bbbfc41c8928ac793R1343 and right now I have PR #299 open for current LOD functions that crash the client if passed an ID of 20000+. You seem to error out if the ID is not between 0 and 20000. Can you give a little more context on why this is? I'm trying to understand what's beyond 19999 that makes it crash the client so I can figure out a permanent solution with good context for the other issue.

@qaisjp commented:
> If there is an underlying bug with `CGameSA::GetModelInfo` then that should be fixed instead.

I'm trying to figure out how to give closure to this 20000 issue.

forkerer commented:
As far as I'm aware, (and in the IDB i'm using), 20000 is original size of that array in gta sa, trying to change physical properties of anything above that crashes the client. I think the additional space reserved in CGameSA (up to 26000) was made either by mistake, or in case of mta team wanting to expand some limits.

patrikjuvonen commented:
Found something old but interesting regarding dynamic objects on the Wiki: https://wiki.multitheftauto.com/wiki/Dynamic_object

sbx320 commented on Client/game_sa/CFxManagerSA.cpp:
```diff
@@ -67,6 +67,32 @@ void CFxManagerSA::OnFxSystemSAInterfaceDestroyed(CFxSystemSAInterface* pFxSyste
+    const char* pChars = sName;
+    CFxSystemBPSAInterface* pFxSystemBlueprint = nullptr;
+
+    _asm
```
Just call the function directly. _asm is usually not needed, breaks optimizations and just causes a lot of clutter.
```suggestion
    using func_t = (CFxSystemBPSAInterface*)(__thiscall*)(CFxManagerSAInterface* pInterface, const char* szName);
    auto func = reinterpret_cast<func_t>(FUNC_FxManager_c__GetSystemByName);
    return func(m_pInterface, sName);
```

sbx320 commented on Client/game_sa/CFxManagerSA.cpp:
```diff
@@ -67,6 +67,32 @@ void CFxManagerSA::OnFxSystemSAInterfaceDestroyed(CFxSystemSAInterface* pFxSyste
+        mov     pFxSystemBlueprint, eax
+    }
+
+    if (!pFxSystemBlueprint)
```
Not needed with the other change, but just for the future:
```cpp
if(!pointer) return nullptr;
return pointer;
```
is the same as 
```
return pointer;
```

sbx320 commented on Client/game_sa/CModelInfoSA.cpp:
```diff
@@ -1367,6 +1368,46 @@ void CModelInfoSA::ResetSupportedUpgrades()
+unsigned short CModelInfoSA::GetObjectPropertiesGroup()
+{
+    unsigned short usGroup = GetInterface()->usDynamicIndex;
+    if (usGroup == 0xFFFF)
```
What is the meaning of a `0xFFFF (-1)` property group?

forkerer replied:
If i remember correctly, it's used by the game to determine that this object id doesn't have any custom physical properties assigned to it, in such cases the default(0) physical properties group is used for most physical calculations on those objects

sbx320 commented on Client/game_sa/CModelInfoSA.cpp:
```diff
@@ -1367,6 +1368,46 @@ void CModelInfoSA::ResetSupportedUpgrades()
+
+void CModelInfoSA::RestoreAllObjectsPropertiesGroups()
+{
+    for (auto& pair : ms_OriginalObjectPropertiesGroups)
```
```suggestion
    for (const auto& pair : ms_OriginalObjectPropertiesGroups)
```

sbx320 commented on Client/game_sa/CObjectGroupPhysicalPropertiesSA.cpp:
```diff
@@ -0,0 +1,369 @@
+    // Make copy of original
+    if (!MapFind(ms_OriginalGroupProperties, m_ucObjectGroup))
+    {
+        auto pOriginalCopy = new CObjectGroupPhysicalPropertiesSAInterface;
```
this should really be a `std::unique_ptr`, as it's rather hard to follow where the memory is freed.

sbx320 commented on Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp:
```diff
@@ -1317,3 +1329,486 @@ int CLuaEngineDefs::EngineResetSurfaceProperties(lua_State* luaVM)
+
+        if (iNewGroup < 0 || iNewGroup > 159)
+        {
+            argStream.SetCustomError("Expected group ID in range [0-159] at argument 1");
```
doesn't quite line up with the error above (0 - 19999 includes 19999, 0 - 159 excludes 159?

forkerer replied:
Changed the upper check to match the format, and now it check against 19999

sbx320 reviewed (changes requested):
Looks good in general. I added some comments, but other than that it should be good.

Only issue apart from that would be the discussion on changing result on usage errors in scripting #821

Commit:
Merge remote-tracking branch 'multitheftauto/master' into setModelGroupDynamicProperty

Commit:
Changes suggested in pull request review

Calling of internal functions without _asm blocks
Const specifier in range loops where needed
Replaced pointers with std::unique_ptrs
Fixed inconsistency with bad arguments checking in some lua functions

ghost commented:
Nice work! Does the collision properties work for custom objects? I'm planning to make `CColModel` object based rather than model based.

forkerer commented:
Since most of the properties are assigned in object constructor, based on model group ID, it should work without problems on custom models that replace dynamic objects. I haven't tested that tho so can't be sure.

sbx320 reviewed (approved):
Looks good. Just needs a decision on the script error thing...

ghost commented:
Once this PR is merged, modelers can use this to generate breakable models: https://github.com/DK22Pac/v2saconv

forkerer committed:
Merge branch 'master' into setModelGroupDynamicProperty

patrikjuvonen commented on Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp:
```diff
@@ -1328,3 +1340,486 @@ int CLuaEngineDefs::EngineResetSurfaceProperties(lua_State* luaVM)
+    CScriptArgReader argStream(luaVM);
+    argStream.ReadNumber(iModelID);
+
+    if (!argStream.HasErrors())
```
These errors should be changed accordingly (https://github.com/multitheftauto/mtasa-blue/issues/821#issuecomment-522187565).

patrikjuvonen commented on Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp:
```diff
@@ -1328,3 +1340,486 @@ int CLuaEngineDefs::EngineResetSurfaceProperties(lua_State* luaVM)
+    argStream.ReadNumber(iModelID);
+    argStream.ReadNumber(iNewGroup);
+
+    if (!argStream.HasErrors())
```
These errors should be changed accordingly (https://github.com/multitheftauto/mtasa-blue/issues/821#issuecomment-522187565).

patrikjuvonen commented on Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp:
```diff
@@ -1328,3 +1340,486 @@ int CLuaEngineDefs::EngineResetSurfaceProperties(lua_State* luaVM)
+    CScriptArgReader argStream(luaVM);
+    argStream.ReadNumber(iModelID);
+
+    if (!argStream.HasErrors())
```
These errors should be changed accordingly (https://github.com/multitheftauto/mtasa-blue/issues/821#issuecomment-522187565).

patrikjuvonen commented on Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp:
```diff
@@ -1328,3 +1340,486 @@ int CLuaEngineDefs::EngineResetSurfaceProperties(lua_State* luaVM)
+    argStream.ReadNumber(iGivenGroup);
+    argStream.ReadEnumString(eProperty);
+
+    if (argStream.HasErrors())
```
These errors should be changed accordingly (https://github.com/multitheftauto/mtasa-blue/issues/821#issuecomment-522187565).

patrikjuvonen commented on Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp:
```diff
@@ -1328,3 +1340,486 @@ int CLuaEngineDefs::EngineResetSurfaceProperties(lua_State* luaVM)
+    argStream.ReadNumber(iGivenGroup);
+    argStream.ReadEnumString(eProperty);
+
+    if (argStream.HasErrors())
```
These errors should be changed accordingly (https://github.com/multitheftauto/mtasa-blue/issues/821#issuecomment-522187565).

patrikjuvonen commented on Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp:
```diff
@@ -1328,3 +1340,486 @@ int CLuaEngineDefs::EngineResetSurfaceProperties(lua_State* luaVM)
+    CScriptArgReader argStream(luaVM);
+    argStream.ReadNumber(iGivenGroup);
+
+    if (argStream.HasErrors())
```
These errors should be changed accordingly (https://github.com/multitheftauto/mtasa-blue/issues/821#issuecomment-522187565).

patrikjuvonen reviewed (changes requested):
These errors should be changed accordingly (https://github.com/multitheftauto/mtasa-blue/issues/821#issuecomment-522187565).

forkerer committed:
Merge remote-tracking branch 'multitheftauto/master' into setModelGroupDynamicProperty

forkerer committed:
Change warnings to errors

forkerer commented:
@patrikjuvonen Done :)

patrikjuvonen commented:
Great job! 👍 Thank you for the PR =)
